File: packages/flutter-im-ui/test/message_projection_test.dart
import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

// Lifecycle mutation and reactions as the thread renders them; fixtures only.

Widget _host(Widget child) => MaterialApp(home: Scaffold(body: child));

const _strings = FlareStrings();
const _recalled = FlareMessageLifecycle(
  mutation: FlareMessageMutationState.recalled,
);

FlareMessageData _message(
  String id,
  String sender,
  String text, {
  FlareMessageLifecycle? lifecycle,
  List<FlareReactionGroup> reactions = const [],
}) => FlareMessageData(
  id: id,
  senderId: sender,
  senderName: sender == 'me' ? 'Me' : 'Ann',
  content: FlareTextContent(text),
  lifecycle: lifecycle,
  reactions: reactions,
);

void main() {
  group('recalled messages', () {
    testWidgets('show a notice instead of their content and take no actions', (
      tester,
    ) async {
      final pressed = <String>[];
      await tester.pumpWidget(
        _host(
          FlareMessageList(
            currentUserId: 'me',
            conversationKind: FlareConversationKind.group,
            messages: [
              _message('1', 'ann', 'secret A', lifecycle: _recalled),
              _message('2', 'me', 'secret B', lifecycle: _recalled),
              _message('3', 'ann', 'still here'),
            ],
            onMessageLongPress: (message) => pressed.add(message.id),
            onSwipeReply: (_) {},
          ),
        ),
      );
      expect(find.textContaining('secret'), findsNothing);
      expect(_strings.messageRecalledGroupOther('Ann'), contains('Ann'));
      final groupOther = find.text(_strings.messageRecalledGroupOther('Ann'));
      expect(groupOther, findsOneWidget);
      expect(find.text(_strings.messageRecalledSelf), findsOneWidget);

      expect(find.byKey(const ValueKey('swipe-reply-1')), findsNothing);
      expect(find.byKey(const ValueKey('swipe-reply-2')), findsNothing);
      expect(find.byKey(const ValueKey('swipe-reply-3')), findsOneWidget);
      await tester.longPress(groupOther);
      await tester.longPress(find.text(_strings.messageRecalledSelf));
      await tester.longPress(find.text('still here'));
      expect(pressed, ['3']);
    });

    testWidgets('read as the peer in a single chat, without status, avatar '
        'or selection', (tester) async {
      final toggled = <String>[];
      await tester.pumpWidget(
        _host(
          Column(
            children: [
              FlareMessageBubble(
                message: _message('1', 'ann', 'secret A', lifecycle: _recalled),
                currentUserId: 'me',
                rowPresentation: const FlareMessageRowPresentation(
                  showAvatar: true,
                  reserveAvatarSpace: true,
                ),
                multiSelectMode: true,
                onToggleSelect: (message) => toggled.add(message.id),
              ),
              FlareMessageBubble(
                message: _message('2', 'me', 'secret B', lifecycle: _recalled),
                currentUserId: 'me',
                multiSelectMode: true,
                onToggleSelect: (message) => toggled.add(message.id),
              ),
            ],
          ),
        ),
      );
      expect(find.textContaining('secret'), findsNothing);
      expect(find.text(_strings.messageRecalledPeer), findsOneWidget);
      expect(find.text(_strings.messageRecalledSelf), findsOneWidget);
      expect(find.byType(FlareAvatar), findsNothing);
      expect(find.byType(FlareMessageStatus), findsNothing);
      expect(find.byIcon(Icons.radio_button_unchecked), findsNothing);
      await tester.tap(find.text(_strings.messageRecalledPeer));
      expect(toggled, isEmpty);
    });
  });

  group('reactions', () {
    const reactions = [
      FlareReactionGroup(emoji: '👍', count: 2, reactedBySelf: true),
      FlareReactionGroup(emoji: '🎉', count: 1),
    ];

    testWidgets('pills toggle the current user reaction', (tester) async {
      final toggled = <(String, String)>[];
      await tester.pumpWidget(
        _host(
          FlareMessageList(
            currentUserId: 'me',
            messages: [
              _message('1', 'ann', 'shipped', reactions: reactions),
              _message(
                '2',
                'me',
                'got it',
                reactions: const [FlareReactionGroup(emoji: '❤️', count: 1)],
              ),
            ],
            onReact: (message, emoji) => toggled.add((message.id, emoji)),
          ),
        ),
      );
      await tester.tap(find.text('🎉 1'));
      await tester.tap(find.text('❤️ 1'));
      expect(toggled, [('1', '🎉'), ('2', '❤️')]);
    });

    testWidgets('pills are display-only without a handler', (tester) async {
      await tester.pumpWidget(
        _host(
          FlareMessageList(
            currentUserId: 'me',
            messages: [_message('1', 'ann', 'shipped', reactions: reactions)],
          ),
        ),
      );
      expect(find.text('👍 2'), findsOneWidget);
      // Labels, not disabled buttons: nothing to announce as a control.
      expect(find.byType(TextButton), findsNothing);
    });

    testWidgets('pills do not toggle in multi-select mode', (tester) async {
      final toggled = <String>[];
      await tester.pumpWidget(
        _host(
          FlareMessageList(
            currentUserId: 'me',
            multiSelectMode: true,
            messages: [_message('1', 'ann', 'shipped', reactions: reactions)],
            onReact: (message, emoji) => toggled.add(emoji),
          ),
        ),
      );
      expect(find.byType(TextButton), findsNothing);
      expect(toggled, isEmpty);
    });
  });
}

File: packages/android-im-ui/src/androidTest/kotlin/com/flare/im/ui/MessageProjectionTest.kt
package com.flare.im.ui

import androidx.compose.material3.MaterialTheme
import androidx.compose.ui.test.assertCountEquals
import androidx.compose.ui.test.hasClickAction
import androidx.compose.ui.test.junit4.createComposeRule
import androidx.compose.ui.test.onAllNodesWithText
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.compose.ui.test.performTouchInput
import androidx.compose.ui.test.longClick
import org.junit.Assert.assertEquals
import org.junit.Rule
import org.junit.Test

/** Lifecycle mutation and reactions as the thread renders them; fixtures only. */
class MessageProjectionTest {
    @get:Rule val compose = createComposeRule()

    private val recalled = FlareMessageLifecycle(mutation = FlareMessageMutationState.Recalled)

    private fun message(id: String, sender: String, text: String, lifecycle: FlareMessageLifecycle? = null, reactions: List<ReactionGroup> = emptyList()) =
        FlareMessageData(id = id, senderId = sender, senderName = if (sender == "me") "我" else "Ann", content = FlareTextContent(text), lifecycle = lifecycle, reactions = reactions)

    @Test fun recalledMessagesShowANoticeInsteadOfTheirContent() {
        val pressed = mutableListOf<String>()
        compose.setContent { MaterialTheme {
            MessageList(
                messages = listOf(
                    message("1", "ann", "私密内容 A", recalled),
                    message("2", "me", "私密内容 B", recalled),
                    message("3", "ann", "还在的消息"),
                ),
                currentUserId = "me",
                conversationKind = FlareConversationKind.Group,
                onMessageLongPress = { pressed += it.id },
            )
        } }
        compose.onAllNodesWithText("私密内容", substring = true).assertCountEquals(0)
        compose.onNodeWithText("Ann 撤回了一条消息").assertExists()
        compose.onNodeWithText("你撤回了一条消息").performTouchInput { longClick() }
        compose.onNodeWithText("还在的消息").performTouchInput { longClick() }
        compose.runOnIdle { assertEquals(listOf("3"), pressed) }
    }

    private val reactions = listOf(ReactionGroup("👍", 2, reactedBySelf = true), ReactionGroup("🎉", 1))

    @Test fun reactionPillsToggleTheCurrentUsersReaction() {
        val toggled = mutableListOf<Pair<String, String>>()
        compose.setContent { MaterialTheme {
            MessageList(
                messages = listOf(message("1", "ann", "上线了", reactions = reactions), message("2", "me", "收到", reactions = listOf(ReactionGroup("❤️", 1)))),
                currentUserId = "me",
                onReact = { msg, emoji -> toggled += msg.id to emoji },
            )
        } }
        compose.onNodeWithText("🎉").performClick()
        compose.onNodeWithText("❤️").performClick()
        compose.runOnIdle { assertEquals(listOf("1" to "🎉", "2" to "❤️"), toggled) }
    }

    @Test fun reactionPillsAreDisplayOnlyWithoutAHandler() {
        compose.setContent { MaterialTheme {
            MessageList(messages = listOf(message("1", "ann", "上线了", reactions = reactions)), currentUserId = "me")
        } }
        compose.onNodeWithText("👍").assertExists()
        compose.onAllNodes(hasClickAction()).assertCountEquals(0)
    }
}
